Doc/library/asyncio-pools.rst

```diff
@@ -0,0 +1,95 @@
+.. currentmodule:: asyncio
+
+.. versionadded:: 3.9
+
+.. _asyncio-pools:
+
+=====
+Pools
+=====
+
+**Source code:** :source:`Lib/asyncio/pools.py`
+
+-----------------------------------------------
+
+.. note::
+    This section of the documentation and all of its members have been
+    added *provisionally* to asyncio's API. For more details, see
+    :term:`provisional api`.
+
+Asyncio pools are high-level, asynchronous context managers that can be used
+to concurrently run blocking functions and methods.
+
+There are many potential use cases, but a particularly useful one is for
+combining libraries without asyncio support with existing asyncio programs.
+Normally, calling a non-async function within the event loop would will result
+in blocking the event loop until the function returns. However, by using a
+pool to run the function, it can be executed in a separate worker (such as a
+thread or process) without blocking the event loop.
+
+.. class:: ThreadPool(concurrency=None)
+
+   An asynchronous threadpool that provides methods to concurrently
+   run IO-bound functions, without blocking the event loop.
+
+   *concurrency* is an optional argument that limits the number of
+   threads to utilize in the threadpool. With the default value of
+   ``None``, the amount of threads used will scale based on the
+   number of processors.
+
+   .. coroutinemethod:: run(/, func, *args, **kwargs)
+
+      Asynchronously run *func* with its arguments and keyword-arguments
+      within the threadpool, and return a :class:`asyncio.Future` object
+      that represents the eventual result of its execution. ::
+
+          async with asyncio.ThreadPool() as pool:
+              await pool.run(time.sleep, 1)
+
+      Raises a :exc:`RuntimeError` if the threadpool is *not* running.
+
+   .. coroutinemethod:: astart()
+
+      Schedule the start of the threadpool and spawn its threads. Note that
+      this function is called automatically when using ``asyncio.ThreadPool``
+      as an asynchronous context manager, and does not need to be called
+      directly.
+
+      Raises a :exc:`RuntimeError` if the threadpool is already running or
+      if it's been closed.
+
+   .. coroutinemethod:: aclose()
```

Comment:
The naming of `astart` and `aclose` don't seem too clear to me. Would it be more descriptive to name them `start` and `shutdown` (like the concurrent.futures.Executor) or maybe `close` (like multiprocess.pool.Pool)?

Reply:
The reasoning behind using `astart` and `aclose` is to clearly communicate that the methods are non-blocking and coroutines that should be awaited. As `start` and `close` are common names in similar APIs, it would be easy to forget that. It also fits quite nicely with the corresponding dunder/system method names of `__aenter__` and `__aexit__`, which is used for async context managers. In most situations though, users should not have to use `astart` and `aclose` explicitly; so I don't think it's an overly significant concern.